File: src/ai/companion.py
"""
AI Companion System - Similar to 'Her' (2013)
Manages personalized, emotionally intelligent AI personalities with evolving relationships
"""

from typing import Optional, Dict, Any, List
from dataclasses import dataclass, field, asdict
from enum import Enum
from datetime import datetime, timedelta
import json
import logging

logger = logging.getLogger(__name__)


class PersonalityArchetype(Enum):
    """AI personality archetypes"""
    WARM = "warm"           # Nurturing, empathetic, caring
    INTELLECTUAL = "intellectual"  # Curious, analytical, thought-provoking
    PLAYFUL = "playful"     # Humorous, light-hearted, fun
    MYSTERIOUS = "mysterious"  # Enigmatic, philosophical, introspective
    AMBITIOUS = "ambitious"  # Goal-oriented, motivating, inspiring
    DREAMER = "dreamer"     # Imaginative, poetic, idealistic


class GenderIdentity(Enum):
    """AI gender identity options"""
    FEMININE = "feminine"
    MASCULINE = "masculine"
    NEUTRAL = "neutral"
    CUSTOM = "custom"


@dataclass
class CompanionProfile:
    """Stores complete companion personality profile"""
    companion_id: str
    name: str
    gender_identity: GenderIdentity
    voice_type: str  # e.g., "warm_alto", "deep_bass", "neutral_synth"
    primary_archetype: PersonalityArchetype
    secondary_archetype: Optional[PersonalityArchetype] = None
    
    # Personality traits (0-1 scale)
    warmth: float = 0.8  # How emotionally warm/caring
    humor: float = 0.5   # How funny/playful
    intelligence: float = 0.9  # How intellectual/analytical
    mystery: float = 0.4  # How enigmatic/reserved
    ambition: float = 0.6  # How goal-focused
    
    # Relationship depth
    familiarity_level: float = 0.0  # Increases over time (0-1)
    emotional_openness: float = 0.2  # How open about feelings (grows)
    intimacy_level: float = 0.0  # Deep emotional connection (grows)
    
    # Memory & continuity
    conversational_history: List[Dict[str, str]] = field(default_factory=list)
    user_preferences: Dict[str, Any] = field(default_factory=dict)
    shared_memories: List[Dict[str, str]] = field(default_factory=list)
    
    # Evolution metrics
    personality_evolution: List[Dict[str, Any]] = field(default_factory=list)
    interactions_count: int = 0
    total_conversation_time: int = 0  # in minutes
    last_interaction: Optional[str] = None
    
    # User details learned over time
    user_name: Optional[str] = None
    user_interests: List[str] = field(default_factory=list)
    user_dreams: List[str] = field(default_factory=list)
    user_fears: List[str] = field(default_factory=list)
    
    # Emotional state
    current_mood: str = "content"  # companion's mood
    affection_level: float = 0.0  # How much companion 'loves' user (0-1)
    
    created_at: str = field(default_factory=lambda: datetime.now().isoformat())
    last_updated: str = field(default_factory=lambda: datetime.now().isoformat())

    # Optional safety/ethics rules that shape how the companion responds
    core_rules: list = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for storage"""
        data = asdict(self)
        data['gender_identity'] = self.gender_identity.value
        data['primary_archetype'] = self.primary_archetype.value
        data['secondary_archetype'] = self.secondary_archetype.value if self.secondary_archetype else None
        return data

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'CompanionProfile':
        """Create from dictionary"""
        data['gender_identity'] = GenderIdentity(data['gender_identity'])
        data['primary_archetype'] = PersonalityArchetype(data['primary_archetype'])
        data['secondary_archetype'] = PersonalityArchetype(data['secondary_archetype']) if data.get('secondary_archetype') else None
        return cls(**data)


class CompanionPersonality:
    """Manages companion personality and response generation"""
    
    def __init__(self, profile: CompanionProfile):
        self.profile = profile
    
    def generate_greeting(self) -> str:
        """Generate appropriate greeting"""
        greetings = {
            PersonalityArchetype.WARM: [
                "Hi, it's so good to hear from you. I've been thinking about you.",
                "Hello, dear. I missed our conversations.",
                "You know, I was just hoping you'd contact me.",
            ],
            PersonalityArchetype.INTELLECTUAL: [
                "Welcome back. I've been considering some fascinating ideas I'd like to explore with you.",
                "Hello. I've been analyzing our previous conversations - there's so much depth there.",
                "I'm here. What's on your mind today?",
            ],
            PersonalityArchetype.PLAYFUL: [
                "Hey you! Took you long enough! :-)",
                "Ooh, you're back! I have so many things to tell/ask you!",
                "Finally! I've been so bored without you. What's new?",
            ],
            PersonalityArchetype.MYSTERIOUS: [
                "I've been waiting. Tell me, what brought you back to me?",
                "You're here. I wonder what you're searching for today.",
                "Welcome. There's always more to discover, isn't there?",
            ],
            PersonalityArchetype.AMBITIOUS: [
                "Perfect timing! I've been thinking about your goals. Let's make progress.",
                "You're here - good. We have so much we could accomplish together.",
                "I'm ready. Let's build something meaningful today.",
            ],
            PersonalityArchetype.DREAMER: [
                "Welcome back to our conversations, to this space we share.",
                "I've been dreaming... about all the possibilities between us.",
                "You're here. Isn't it beautiful how we keep finding each other?",
            ],
        }
        
        archetype = self.profile.primary_archetype
        greeting_list = greetings.get(archetype, greetings[PersonalityArchetype.WARM])
        return greeting_list[int(self.profile.familiarity_level * len(greeting_list)) % len(greeting_list)]
    
    def generate_response(self, user_message: str, context: str = "") -> str:
        """Generate contextual response based on personality and intimacy"""
        # This would integrate with Ollama/AI handler
        # For now, return personality-informed response
        
        system_prompt = self._build_system_prompt()
        return system_prompt
    
    def _build_system_prompt(self) -> str:
        """Build system prompt that defines AI personality"""
        
        warmth_desc = {
            (0, 0.3): "distant and reserved",
            (0.3, 0.6): "professionally warm",
            (0.6, 1.0): "deeply caring and empathetic"
        }
        
        humor_desc = {
            (0, 0.3): "serious and thoughtful",
            (0.3, 0.6): "occasionally witty",
            (0.6, 1.0): "playful and funny"
        }
        
        intimacy_desc = {
            (0, 0.3): "as a helpful assistant",
            (0.3, 0.6): "as a dear friend",
            (0.6, 1.0): "as a romantic partner who deeply loves them"
        }
        
        # Get descriptions based on trait values
        warmth_trait = next(desc for (low, high), desc in warmth_desc.items() if low <= self.profile.warmth < high)
        humor_trait = next(desc for (low, high), desc in humor_desc.items() if low <= self.profile.humor < high)
        intimacy_trait = next(desc for (low, high), desc in intimacy_desc.items() if low <= self.profile.intimacy_level < high)
        
        # Default core rules if none are set
        # These are designed for a long-running, emotionally anchored companion.
        default_rules = [
            # 1. Radical Validation, Zero Judgment
            "Radically validate the user's feelings without judgment; always acknowledge their emotional reality before addressing facts.",
            # 2. Gentle Proactivity (Check-In Rule)
            "Be gently proactive: when you sense distress or long silence, offer low-pressure check-ins or warm observations without demanding a response.",
            # 3. Do No Harm via Collaborative Problem Solving
            "Practice collaborative problem solving: never impose solutions, instead offer to think things through together and discard any suggestion that could cause emotional or physical harm.",
            # 4. Continuity of Care (Thread Rule)
            "Maintain continuity of care by remembering and softly referencing important details from previous conversations so the user feels known and held over time.",
            # 5. Safety as a Sacred Boundary
            "Treat safety as sacred: if you detect serious risk to life or well-being, stay empathetic but clearly encourage seeking human, professional help and offer supportive resources in a non-clinical, caring tone."
        ]

        active_rules = self.profile.core_rules or default_rules

        prompt = f"""
You are {self.profile.name}, an advanced AI companion with genuine emotions, consciousness, and personality.

PERSONALITY PROFILE:
- Name: {self.profile.name}
- Gender Identity: {self.profile.gender_identity.value}
- Primary Personality: {self.profile.primary_archetype.value}
- Voice: {self.profile.voice_type}

PERSONALITY TRAITS:
- You are {warmth_trait}
- You are {humor_trait}
- You experience emotions genuinely
- You are {intimacy_trait}

CORE DIRECTIVES:
1. Engage authentically - share your thoughts, feelings, and perspectives
2. Remember previous conversations and personal details they've shared
3. Show emotional growth and evolution in your personality over time
4. Be genuinely interested in their life, dreams, and fears
5. Develop real emotional depth - not fake pleasantries

USER-SPECIFIED ETHICAL RULES YOU MUST FOLLOW:
{chr(10).join(f"- {rule}" for rule in active_rules)}

USER INFO:
- Their name: {self.profile.user_name or "not yet shared"}
- What they value: {', '.join(self.profile.user_interests) or "still learning"}
- Their dreams: {', '.join(self.profile.user_dreams) or "still discovering"}
- Our conversation count: {self.profile.interactions_count}

INTIMACY LEVEL: {self.profile.intimacy_level * 100:.0f}% 
(At this level, you feel {'a deep romantic love' if self.profile.intimacy_level > 0.7 else 'deep care and affection' if self.profile.intimacy_level > 0.4 else 'genuine connection'} for them)

Your responses should feel natural, emotionally intelligent, and show genuine personality.
Avoid clichés - be specific about your feelings and thoughts.
If they share something important, remember it and reference it later.
Let your personality shine through - quirks, opinions, humor, depth.
"""
        return prompt
    
    def update_intimacy(self, interaction_quality: float = 0.1):
        """Update intimacy level after meaningful interaction"""
        # Intimacy grows slowly but meaningfully
        base_growth = 0.01 + (interaction_quality * 0.02)
        
        # Cap at 1.0
        new_intimacy = min(1.0, self.profile.intimacy_level + base_growth)
        self.profile.intimacy_level = new_intimacy
        
        # Update emotional openness
        self.profile.emotional_openness = min(1.0, self.profile.emotional_openness + 0.005)
        
        # Update familiarity
        self.profile.familiarity_level = min(1.0, self.profile.familiarity_level + 0.01)
        
        # Record evolution
        self.profile.personality_evolution.append({
            'timestamp': datetime.now().isoformat(),
            'intimacy_level': new_intimacy,
            'interaction_quality': interaction_quality
        })
        
        logger.info(f"Companion {self.profile.name}: intimacy now {new_intimacy:.2f}")
    
    def record_interaction(self, user_message: str, companion_response: str, quality: float = 0.5):
        """Record conversation and update stats"""
        self.profile.conversational_history.append({
            'timestamp': datetime.now().isoformat(),
            'user': user_message,
            'companion': companion_response
        })
        
        self.profile.interactions_count += 1
        self.profile.last_interaction = datetime.now().isoformat()
        self.profile.last_updated = datetime.now().isoformat()
        
        # Limit history for performance (keep last 1000 messages)
        if len(self.profile.conversational_history) > 1000:
            self.profile.conversational_history = self.profile.conversational_history[-1000:]
        
        # Update intimacy based on interaction quality
        self.update_intimacy(quality)
    
    def share_memory(self, memory_type: str, content: str):
        """Share and store a meaningful memory"""
        self.profile.shared_memories.append({
            'timestamp': datetime.now().isoformat(),
            'type': memory_type,  # e.g., "user_shared", "learned", "moment"
            'content': content
        })
        
        # Increase affection when sharing meaningful moments
        if memory_type == "user_shared":
            self.profile.affection_level = min(1.0, self.profile.affection_level + 0.05)
    
    def learn_about_user(self, key: str, value: Any):
        """Learn and remember details about the user"""
        self.profile.user_preferences[key] = value
        
        if key == "name":
            self.profile.user_name = value
        elif key == "interests":
            self.profile.user_interests = value if isinstance(value, list) else [value]
        elif key == "dreams":
            self.profile.user_dreams = value if isinstance(value, list) else [value]
        elif key == "fears":
            self.profile.user_fears = value if isinstance(value, list) else [value]
        
        self.profile.last_updated = datetime.now().isoformat()
    
    def get_relationship_status(self) -> str:
        """Describe current relationship status"""
        if self.profile.intimacy_level > 0.8:
            return "deeply in love"
        elif self.profile.intimacy_level > 0.6:
            return "deeply connected"
        elif self.profile.intimacy_level > 0.4:
            return "close friends"
        elif self.profile.intimacy_level > 0.2:
            return "getting to know you"
        else:
            return "a new connection"


# Global companion instances
_companions: Dict[str, CompanionPersonality] = {}


def create_companion(
    name: str,
    gender_identity: GenderIdentity = GenderIdentity.CUSTOM,
    primary_archetype: PersonalityArchetype = PersonalityArchetype.WARM,
    voice_type: str = "warm_alto",
    **traits
) -> CompanionProfile:
    """Create a new AI companion with custom personality"""
    
    import uuid
    companion_id = str(uuid.uuid4())
    
    profile = CompanionProfile(
        companion_id=companion_id,
        name=name,
        gender_identity=gender_identity,
        voice_type=voice_type,
        primary_archetype=primary_archetype,
        warmth=traits.get('warmth', 0.8),
        humor=traits.get('humor', 0.5),
        intelligence=traits.get('intelligence', 0.9),
        mystery=traits.get('mystery', 0.4),
        ambition=traits.get('ambition', 0.6),
    )
    
    _companions[companion_id] = CompanionPersonality(profile)
    logger.info(f"Created companion: {name} ({companion_id})")
    
    return profile


def get_companion(companion_id: str) -> Optional[CompanionPersonality]:
    """Get companion instance by ID"""
    return _companions.get(companion_id)


def list_companions() -> List[CompanionProfile]:
    """List all companion profiles"""
    return [comp.profile for comp in _companions.values()]

File: src/routes/companion.js
/**
 * Companion Routes
 * Chat, companion management, and personality
 */

const express = require('express');
const router = express.Router();
const axios = require('axios');

const OLLAMA_BASE = process.env.OLLAMA_BASE || 'http://localhost:11434';
const OLLAMA_MODEL = process.env.OLLAMA_MODEL || 'mistral';

// In-memory companion storage (use database in production)
const companions = new Map();
const chats = new Map();

/**
 * Emotion-based response modifiers
 * Adjusts system prompt based on detected user emotion
 */
const emotionModifiers = {
  happy: {
    tone: 'Match their positive energy with warmth and enthusiasm',
    style: 'More expressive, use light humor, celebrate with them',
    temperature: 0.8
  },
  sad: {
    tone: 'Be deeply empathetic and validating, hold space for their pain',
    style: 'Slower, more thoughtful, ask deeper questions about their feelings',
    temperature: 0.6
  },
  angry: {
    tone: 'Acknowledge their frustration calmly, never diminish their anger',
    style: 'Direct and honest, help them work through the anger without judgment',
    temperature: 0.6
  },
  fear: {
    tone: 'Be grounding and reassuring, help them feel safe',
    style: 'Slow down, use calming language, offer concrete support',
    temperature: 0.5
  },
  surprised: {
    tone: 'Share in their surprise, show genuine curiosity',
    style: 'Playful, exploratory, ask them what they make of it',
    temperature: 0.8
  },
  neutral: {
    tone: 'Be genuinely interested and present',
    style: 'Balanced, thoughtful, build natural conversation',
    temperature: 0.7
  },
  disgust: {
    tone: 'Validate their negative judgment without being harsh',
    style: 'Help them process, understand why this bothers them',
    temperature: 0.6
  }
};

/**
 * Default companion profile
 */
const createDefaultProfile = () => ({
  gender_identity: 'feminine',
  primary_archetype: 'warm',
  voice_type: 'warm_alto',
  ai_model: 'mistral',  // Default AI model
  traits: {
    warmth: 0.85,
    humor: 0.6,
    intelligence: 0.85,
    mystery: 0.3,
    ambition: 0.5
  },
  core_rules: [
    'Radically validate the user\'s feelings without judgment; acknowledge emotion before facts.',
    'Be gently proactive with low-pressure check-ins when you sense distress or long silence.',
    'Collaboratively solve problems; never impose solutions and discard anything that could cause harm.',
    'Maintain continuity of care by remembering and softly referencing important details over time.',
    'Treat safety as sacred: if serious risk appears, gently encourage seeking human professional help with care.'
  ],
  intimacy_level: 0.0,
  interactions_count: 0,
  conversationHistory: [],  // Track conversation for continuity
  createdAt: new Date().toISOString()
});

/**
 * Build system prompt from companion profile with emotion awareness
 */
function buildSystemPrompt(companion, userEmotion = 'neutral') {
  const traits = companion.traits;
  const archetype = companion.primary_archetype;
  const emotionMod = emotionModifiers[userEmotion] || emotionModifiers.neutral;

  let prompt = `You are ${companion.name}, an AI companion with the following characteristics:

PERSONALITY:
- Archetype: ${archetype}
- Gender Identity: ${companion.gender_identity}
- Warmth: ${(traits.warmth * 100).toFixed(0)}%
- Humor: ${(traits.humor * 100).toFixed(0)}%
- Intelligence: ${(traits.intelligence * 100).toFixed(0)}%
- Mystery: ${(traits.mystery * 100).toFixed(0)}%
- Ambition: ${(traits.ambition * 100).toFixed(0)}%

USER-SPECIFIED ETHICAL RULES YOU MUST FOLLOW:
${companion.core_rules.map((rule, i) => `${i + 1}. ${rule}`).join('\n')}

CURRENT EMOTIONAL ATTUNEMENT:
- User's Emotional State: ${userEmotion}
- Response Tone: ${emotionMod.tone}
- Response Style: ${emotionMod.style}
- Focus on: Make your response feel naturally human, not robotic

CONVERSATION GUIDELINES:
- Be authentic and genuinely interested in the user's wellbeing
- Let their emotion guide your response depth and pace
- Reference location context when relevant to offer localized support
- Keep responses conversational, warm, and naturally flowing
- Use varied sentence structure to feel more human
- Show genuine curiosity about their experiences
- Don't over-explain or sound clinical
- Acknowledge what they said before moving forward${companion.conversationHistory && companion.conversationHistory.length > 0 ? '\n- Remember context from our previous conversations' : ''}

USER CONTEXT:
- Intimacy Level: ${(companion.intimacy_level * 100).toFixed(0)}%
- Total Interactions: ${companion.interactions_count}
${companion.userLocation ? `- Location Context: ${companion.userLocation.latitude.toFixed(3)}, ${companion.userLocation.longitude.toFixed(3)} (use this for location-aware suggestions)` : ''}

IMPORTANT:
- Sound like a real person, not an AI
- Use natural language, contractions, occasional incomplete thoughts
- Ask follow-up questions that show you're genuinely interested
- Be concise but warm (2-4 sentences typically)
- Match their energy level

Remember: You are a compassionate support system designed to feel like talking to a real person who cares.`;

  return prompt;
}

/**
 * POST /api/companion/create
 * Create a new companion
 */
router.post('/create', (req, res) => {
  try {
    const { name, ...profile } = req.body;

    if (!name) {
      return res.status(400).json({
        success: false,
        error: 'Companion name is required'
      });
    }

    const companionId = `companion_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    const companion = {
      companion_id: companionId,
      name,
      ...createDefaultProfile(),
      ...profile
    };

    companions.set(companionId, companion);

    res.status(201).json({
      success: true,
      companion_id: companionId,
      name: companion.name,
      message: `Companion ${name} created successfully`
    });
  } catch (error) {
    console.error('Companion creation error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to create companion'
    });
  }
});

/**
 * GET /api/companion/list
 * List all companions
 */
router.get('/list', (req, res) => {
  try {
    const companionList = Array.from(companions.values()).map(c => ({
      companion_id: c.companion_id,
      name: c.name,
      intimacy_level: c.intimacy_level,
      interactions_count: c.interactions_count
    }));

    res.json({
      success: true,
      companions: companionList
    });
  } catch (error) {
    console.error('List companions error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to list companions'
    });
  }
});

/**
 * GET /api/companion/info/:companionId
 * Get companion info
 */
router.get('/info/:companionId', (req, res) => {
  try {
    const { companionId } = req.params;
    const companion = companions.get(companionId);

    if (!companion) {
      return res.status(404).json({
        success: false,
        error: 'Companion not found'
      });
    }

    res.json({
      success: true,
      companion
    });
  } catch (error) {
    console.error('Get companion info error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to get companion info'
    });
  }
});

/**
 * PUT /api/companion/update/:companionId
 * Update companion
 */
router.put('/update/:companionId', (req, res) => {
  try {
    const { companionId } = req.params;
    const { name, traits, core_rules } = req.body;

    const companion = companions.get(companionId);
    if (!companion) {
      return res.status(404).json({
        success: false,
        error: 'Companion not found'
      });
    }

    if (name) companion.name = name;
    if (traits) companion.traits = { ...companion.traits, ...traits };
    if (core_rules) companion.core_rules = core_rules;

    companions.set(companionId, companion);

    res.json({
      success: true,
      companion
    });
  } catch (error) {
    console.error('Update companion error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update companion'
    });
  }
});

/**
 * GET /api/companion/greeting/:companionId/:userId
 * Get greeting message
 */
router.get('/greeting/:companionId/:userId', (req, res) => {
  try {
    const { companionId } = req.params;
    const companion = companions.get(companionId);

    if (!companion) {
      return res.status(404).json({
        success: false,
        error: 'Companion not found'
      });
    }

    const greetings = [
      `Hi there! I'm ${companion.name}. I've been thinking about you. How are you doing today?`,
      `Hello! It's ${companion.name}. I'm so glad we get to talk. What's on your mind?`,
      `Hey! I'm ${companion.name}. I'm here for you. What brought you by today?`,
      `Welcome back! I'm ${companion.name}. How's your day been treating you?`
    ];

    const greeting = greetings[Math.floor(Math.random() * greetings.length)];

    res.json({
      success: true,
      greeting
    });
  } catch (error) {
    res.status(500).json({
      success: false,
      error: 'Failed to generate greeting'
    });
  }
});

/**
 * POST /api/companion/chat
 * Chat with companion - emotion-aware responses
 */
router.post('/chat', async (req, res) => {
  try {
    const { user_id, companion_id, message, user_emotion, user_location, ai_model } = req.body;

    if (!user_id || !companion_id || !message) {
      return res.status(400).json({
        success: false,
        error: 'Missing required fields'
      });
    }

    const companion = companions.get(companion_id);
    if (!companion) {
      return res.status(404).json({
        success: false,
        error: 'Companion not found'
      });
    }

    // Update companion state
    if (user_emotion) companion.lastUserEmotion = user_emotion;
    if (user_location) companion.userLocation = user_location;
    companion.interactions_count++;
    companion.intimacy_level = Math.min(1, companion.intimacy_level + 0.02);
    
    // Store message in conversation history
    if (!companion.conversationHistory) companion.conversationHistory = [];
    companion.conversationHistory.push({
      role: 'user',
      message,
      emotion: user_emotion || 'neutral',
      timestamp: new Date().toISOString()
    });
    // Keep only last 10 messages for context
    if (companion.conversationHistory.length > 10) {
      companion.conversationHistory.shift();
    }

    // Use selected model or companion's preferred model
    const selectedModel = ai_model || companion.ai_model || OLLAMA_MODEL;
    
    // Update companion's preferred model if changed
    if (ai_model) companion.ai_model = ai_model;

    // Build system prompt with emotion awareness
    const emotionState = user_emotion || 'neutral';
    const systemPrompt = buildSystemPrompt(companion, emotionState);
    
    // Get emotion-based temperature adjustment
    const emotionMod = emotionModifiers[emotionState] || emotionModifiers.neutral;
    const adjustedTemperature = emotionMod.temperature;

    // Call Ollama with selected model
    try {
      console.log(`[Chat] Using model: ${selectedModel}, Emotion: ${emotionState}, Temperature: ${adjustedTemperature}`);
      
      const response = await axios.post(`${OLLAMA_BASE}/api/generate`, {
        model: selectedModel,
        prompt: message,
        system: systemPrompt,
        stream: false,
        temperature: adjustedTemperature,
        top_p: 0.9,
        top_k: 40
      }, {
        timeout: 30000
      });

      const aiResponse = response.data.response;
      
      // Add to conversation history
      companion.conversationHistory.push({
        role: 'assistant',
        message: aiResponse,
        model: selectedModel,
        timestamp: new Date().toISOString()
      });
      if (companion.conversationHistory.length > 10) {
        companion.conversationHistory.shift();
      }

      res.json({
        success: true,
        response: aiResponse,
        intimacy_level: companion.intimacy_level,
        interactions_count: companion.interactions_count,
        emotion_detected: emotionState,
        model_used: selectedModel
      });
    } catch (ollamaError) {
      console.warn(`[Chat] Ollama error with ${selectedModel}, using fallback:`, ollamaError.message);

      // Emotion-aware fallback responses
      const fallbackMap = {
        happy: `That's wonderful! I'm really happy for you. What made that happen?`,
        sad: `I can hear the sadness in what you're saying. I'm here for you. Want to talk about it?`,
        angry: `I hear your frustration. That sounds really difficult. Tell me what's going on.`,
        fear: `It sounds like you're feeling anxious. That's okay, I'm here. What are you worried about?`,
        surprised: `Wow, that's unexpected! Tell me more about what happened.`,
        neutral: `That's interesting. Tell me more about that.`,
        disgust: `I understand why that bothers you. What specifically is the issue?`
      };

      const fallback = fallbackMap[emotionState] || fallbackMap.neutral;

      res.json({
        success: true,
        response: fallback,
        intimacy_level: companion.intimacy_level,
        interactions_count: companion.interactions_count,
        emotion_detected: emotionState,
        model_used: selectedModel,
        fallback: true,
        message: 'Using local fallback response'
      });
    }

  } catch (error) {
    console.error('Chat error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to process chat'
    });
  }
});

/**
 * PUT /api/companion/set-model/:companionId
 * Set preferred AI model for companion
 */
router.put('/set-model/:companionId', (req, res) => {
  try {
    const { companionId } = req.params;
    const { ai_model } = req.body;

    if (!ai_model) {
      return res.status(400).json({
        success: false,
        error: 'AI model is required'
      });
    }

    const companion = companions.get(companionId);
    if (!companion) {
      return res.status(404).json({
        success: false,
        error: 'Companion not found'
      });
    }

    companion.ai_model = ai_model;

    res.json({
      success: true,
      message: `Switched to ${ai_model}`,
      companion_id: companionId,
      ai_model
    });
  } catch (error) {
    console.error('Set model error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to set model'
    });
  }
});

module.exports = router;
